### ISSUE TITLE
disconnection and reconnection when using fetchone() after EOF (with block=False)

### ISSUE THREAD

**Issue by giacomolozito** [noplay/python-mysql-replication#121]
Hello,
I have a python program that uses python-mysql-replication with block=False (and resume_stream=True) and gets events with fetchone() within a loop.

With block=False, fetchone() returns None when it gets an EOF from PyMySQL (no new event available in binary log).

However the next attempt with fetchone(), if there are still no new events, does not return None (EOF) again, but it causes PyMySQL to trigger an error 2013 ("Lost connection to MySQL server during query") because pymysql _read_packet() will try to read a new 4 bytes header but 0 bytes are available (_read_packet() and _read_bytes() definition from https://github.com/PyMySQL/PyMySQL/blob/1f9222feb2a668c4e918000e5358b98afbde38d7/pymysql/connections.py)

This exception causes python-mysql-replication to assume that the connection is gone, it sets self.__connected_stream = False and re-creates the stream connection upon the next loop iteration in fetchone()
(fetchone() definition in https://github.com/noplay/python-mysql-replication/blob/afe27999ccc9df6d47f9a69064e481cff4798a62/pymysqlreplication/binlogstream.py)

This keeps going with disconnections and reconnections until a new event is available.

Is this the intended behavior?
I would have expected fetchone() to keep returning None until something new is available without re-creating the connection every time, but perhaps this is due to how PyMySQL handles further attempts to read beyond EOF?

On a sidenote, with PyPy 2.6 (based on python 2.7) this behavior is especially troublesome because the faulted connections are not automatically closed (as they are with standard python 2.6.6) and each disconnection/reconnection waiting for a new event causes one more connection to stay in CLOSE_WAIT state forever, until resources starvation occurs.
I fixed this issue with PyPy by setting an explicit close in python-mysql-replication when self.__connected_stream is set False:

```python
            except pymysql.OperationalError as error:
                code, message = error.args
                if code in MYSQL_EXPECTED_ERROR_CODES:
                    self._stream_connection.close() # <--- prevents CLOSE_WAIT socket spam with PyPy
                    self.__connected_stream = False
                    continue
```

Any thoughts about the behavior described above is appreciated.

**Comment by noplay** [comment]
Could you make a PR for the close connection? Thanks a lot

**Comment by noplay** [comment]
You are right you should not lost the connection if you use block False

**Comment by giacomolozito** [comment]
Hello,
had a good look into MySQL 5.6.24 source code.
Turns out that disconnection on EOF when using non-blocking replication (BINLOG_DUMP_NON_BLOCK set) is by design in MySQL.

So I guess that when designing a program and being unwilling to have program execution blocked for a potentially long time (until a new event is available in binlog), there are two ways to go:

- use block=False and let python-mysql-replication deal with the reconnection every time fetchone() is called after an EOF (I added a time.sleep(x) into my program to prevent this from happening too quickly and the close() patch eliminates lingering sockets with PyPy); this is now working just fine for me

- use block=True and deal with blocking I/O in a different way; i.e. putting the binlogreader code into a separate process with multiprocessing, or changing pymysql settimeout from None to something to ensure blocking I/O will still have a chance to return, or probably some other way I haven't thought of...

Still, disconnection is by design so this issue can be closed.
